File: python/triangle/triangle.py
def is_triangle(fn):
    return lambda s: fn(s) and all(s) and 2*max(s) < sum(s)

@is_triangle
def is_equilateral(sides):
    return len(set(sides)) == 1

@is_triangle
def is_isosceles(sides):
    return len(set(sides)) <= 2

@is_triangle
def is_scalene(sides):
    return len(set(sides)) == 3

File: java/triangle/src/main/java/Triangle.java
import java.util.Arrays;
import java.util.List;

class Triangle {
    private double min, mid, max;

    Triangle(double a, double b, double c) throws TriangleException {
        List<Double> sides = Arrays.asList(a, b, c);
        sides.sort(Double::compare);

        min = sides.get(0);
        mid = sides.get(1);
        max = sides.get(2);

        if (min <= 0 || max > min + mid) {
            throw new TriangleException();
        }
    }

    boolean isEquilateral() {
       return min == max;
    }

    boolean isIsosceles() {
       return min == mid || mid == max;
    }

    boolean isScalene() {
       return min != mid && mid != max;
    }
}
